api/APIValidation.ts

```diff
@@ -0,0 +1,75 @@
+import { ZodError, ZodTypeAny, z } from "zod";
+import { addLogHandler, consoleLogHandler, logger } from "../logging";
+import { APIMiddleware, TiFAPIInputContext, TiFAPIResponse } from "./TransportTypes";
+
+const log = logger("tif.api.validation")
+addLogHandler(consoleLogHandler())
+
+const zodValidation = async <T>(data: T, schema: ZodTypeAny): Promise<T | "failure"> => {
+  try {
+      return await schema.parseAsync(data);
+  } catch (e) {
+      if (e instanceof ZodError) {
+          log.error("Zod Schema Error Message", {
+              zodError: JSON.parse(e.message)
+          });
+      } else {
+          log.error(e);
+      }
+      return "failure";
+  }
+};
+
+type ValidationResult = "invalid-request" | "unexpected-response" | "invalid-response" | "passed"
+
+type ValidationResultParser = (status: ValidationResult, value: TiFAPIInputContext<any> | TiFAPIResponse<any>) => (TiFAPIResponse<any>)
```

Comment:
Also would be nice to have typings in which one of the validation results would specify the type of `value`. Maybe a discriminated union is better here?